### ISSUE TITLE
deepspeed zero3 also save ckpt only in rank 0?

### ISSUE THREAD

**Issue by Alex-Ruan** [intelligent-machine-learning/dlrover#1256]
According to /opt/conda/envs/dlrover/lib/python3.10/site-packages/dlrover/trainer/torch/flash_checkpoint/deepspeed_engine.py
```python   
def save_to_storage(self, step, state_dict, paths):
        """
        Asynchonously saves the state dict into the storage. It synchonously
        saves the state dict into the shared memory and put the path
        into a shared queue. The agent in the main process waits for the queue
        for save the state dict in the shared memory into the storage.
        Only rank 0 saves the state dict into the storage.
        Args:
            step (int): the global iteration step.
            state_dict (dict): the state dict of model and optimizer to save.
            paths (dict): the key is a category in
                ["model_states", "optim_states"] of the state dict and
                the value is the path of storage to save.
        """
        success = True
        if step > self._cached_step:
            success = self.save_to_memory(step, state_dict, paths)
        if dist.is_initialized():
            dist.barrier()
        # Only local rank 0 to notify the saving event to the agent.
        if self._local_rank == 0 and success:
            event = CheckpointEvent(type=CheckpointEventType.SAVE, step=step)
            self._event_queue.put(event)
        return success
```

`self._local_rank == 0 ` means it saves ckpt only in rank0, but I use deepspeed zero3, the model state dict is divided into mutiply gpus.
请问deepspeed zero3的时候模型被shard到多卡了，保存ckpt的时候还是只异步保存rank 0那张卡的ckpt吗，感谢帮忙解答一下！

**Comment by workingloong** [comment]
不是的，这个是只需要 local rank 0 去通知异步进程执行将共享内存的ckpt 导出到磁盘。

**Comment by Alex-Ruan** [comment]
第一，所以如果有4张卡，rank0, 1, 2, 3的`zero_pp_rank_[这里是local rank序号]_mp_rank_00_model_states.pt`和`zero_pp_rank_[这里是local `rank序号]_mp_rank_00_optim_states.pt`都会保存，而不仅仅是如下我运行出来的结果中只有`zero_pp_rank_0_mp_rank_00_model_states.pt`和`zero_pp_rank_0_mp_rank_00_optim_states.pt`

> root@ba00b1c2d4b2:/# ls -lah 14B_4flashckpt_7a100
total 12M
drwxr-xr-x 5 root root 4.0K Sep  4 02:02 .
drwxr-xr-x 1 root root 4.0K Sep  4 02:01 ..
drwxr-xr-x 2 root root 4.0K Sep  4 02:01 ._dlrover_ckpt_stage
-rw-r--r-- 1 root root   80 Sep  4 02:02 added_tokens.json
-rw-r--r-- 1 root root  305 Sep  4 02:02 all_results.json
drwxr-xr-x 3 root root 4.0K Sep  4 01:59 checkpoint-15
drwxr-xr-x 3 root root 4.0K Sep  4 02:01 checkpoint-45
-rw-r--r-- 1 root root  698 Sep  4 02:01 config.json
-rw-r--r-- 1 root root    2 Sep  4 02:01 dlrover_latest.txt
-rw-r--r-- 1 root root  160 Sep  4 02:02 eval_results.json
-rw-r--r-- 1 root root  117 Sep  4 02:01 generation_config.json
-rw-r--r-- 1 root root    2 Sep  4 02:01 latest
-rw-r--r-- 1 root root 1.6M Sep  4 02:02 merges.txt
-rw-r--r-- 1 root root  39K Sep  4 02:02 pytorch_model.bin.index.json
-rw-r--r-- 1 root root  370 Sep  4 02:02 special_tokens_map.json
-rw-r--r-- 1 root root 6.8M Sep  4 02:02 tokenizer.json
-rw-r--r-- 1 root root 1.4K Sep  4 02:02 tokenizer_config.json
-rw-r--r-- 1 root root  165 Sep  4 02:02 train_results.json
-rw-r--r-- 1 root root  716 Sep  4 02:01 trainer_log.jsonl
-rw-r--r-- 1 root root 1.3K Sep  4 02:02 trainer_state.json
-rw-r--r-- 1 root root 5.9K Sep  4 02:02 training_args.bin
-rw-r--r-- 1 root root 2.7M Sep  4 02:02 vocab.json
root@ba00b1c2d4b2:/# ls -lah 14B_4flashckpt_7a100/checkpoint-45
total 27G
drwxr-xr-x 3 root root 4.0K Sep  4 02:01 .
drwxr-xr-x 5 root root 4.0K Sep  4 02:02 ..
drwxr-xr-x 2 root root 4.0K Sep  4 02:01 global_step45
-rw-r--r-- 1 root root 4.6G Sep  4 02:00 pytorch_model-00001-of-00006.bin
-rw-r--r-- 1 root root 4.7G Sep  4 02:00 pytorch_model-00002-of-00006.bin
-rw-r--r-- 1 root root 4.7G Sep  4 02:00 pytorch_model-00003-of-00006.bin
-rw-r--r-- 1 root root 4.7G Sep  4 02:01 pytorch_model-00004-of-00006.bin
-rw-r--r-- 1 root root 4.7G Sep  4 02:01 pytorch_model-00005-of-00006.bin
-rw-r--r-- 1 root root 3.3G Sep  4 02:01 pytorch_model-00006-of-00006.bin
-rw-r--r-- 1 root root  21K Sep  4 02:01 rng_state_0.pth
-rw-r--r-- 1 root root  627 Sep  4 02:01 scheduler.pt
root@ba00b1c2d4b2:/# ls -lah 14B_4flashckpt_7a100/checkpoint-45/global_step45
total 23G
drwxr-xr-x 2 root root 4.0K Sep  4 02:01 .
drwxr-xr-x 3 root root 4.0K Sep  4 02:01 ..
-rw-r--r-- 1 root root 198K Sep  4 02:01 zero_pp_rank_0_mp_rank_00_model_states.pt
-rw-r--r-- 1 root root  23G Sep  4 02:01 zero_pp_rank_0_mp_rank_00_optim_states.pt
---

第二，请问
```python
def check_all_rank_ready(group: dist.ProcessGroup, ready: bool):
    """
    Check whether all ranks are ready.
    """
    if not group and not dist.is_initialized():
        return ready
    backend = dist.get_backend(group)
    local_rank = env_utils.get_local_rank()
    device = "cpu" if backend == "gloo" else f"cuda:{local_rank}"
    rt = ReadyTensor.singleton_instance(device)
    value = 0 if ready else 1
    rt.tensor[0] = value
    dist.all_reduce(rt.tensor, group=group)# 这行导致local rank 0的save_to_storage卡住，注释之后ready==True不卡了
    ready = rt.tensor == 0
    return ready
![img_v2_0adfe6f3-0166-4583-bd9a-0980640148el](https://github.com/user-attachments/assets/3fc3fe54-d75b-4c70-9bd9-0459817d7966)

中ready一直返回False可能是什么原因导致的呢？
